test/common/stats/stats_matcher_test.cc

```diff
@@ -0,0 +1,249 @@
+#include "envoy/config/metrics/v2/stats.pb.h"
+
+#include "common/stats/stats_matcher_impl.h"
+
+#include "test/test_common/utility.h"
+
+#include "gtest/gtest.h"
+
+using testing::IsFalse;
+using testing::IsTrue;
+
+namespace Envoy {
+namespace Stats {
+
+class StatsMatcherTest : public testing::Test {
+public:
+  StatsMatcherTest() {}
+  void SetUp() override { stats_matcher_impl_ = std::make_unique<StatsMatcherImpl>(stats_config_); }
+  void TearDown() override {
+    for (const auto& stat_name : expected_to_pass_) {
+      EXPECT_THAT(stats_matcher_impl_->rejects(stat_name), IsFalse());
+    }
+    for (const auto& stat_name : expected_to_fail_) {
+      EXPECT_THAT(stats_matcher_impl_->rejects(stat_name), IsTrue());
+    }
+  }
+
+  envoy::config::metrics::v2::StatsConfig stats_config_;
+
+  std::vector<std::string> expected_to_pass_;
+  std::vector<std::string> expected_to_fail_;
+
+private:
+  std::unique_ptr<StatsMatcherImpl> stats_matcher_impl_;
+};
+
+TEST_F(StatsMatcherTest, CheckDefault) {
+  // With no set fields, everything should be allowed through.
+  SetUp();
```

Comment:
You don't need explicit Setup() calls here; in fact your SetUp is running twice for each test. You might not have noticed because it's also idempotent...oh I see. You are sometimes calling SetUp after changing some config.

Let's call this SetUp() method something else, and instead maybe pass in the Regex.  E.g.

initRegex(".*");

Some of the more complex cases don't fit that model, in which case just elaborate the customization and construction of the config for those.

Reply:
Not enough of my tests are _just_ regexes, but I found a middle ground which increases readability and doesn't call setup(). See 8bb4aa9a0.